meson: add luajit as an alternative option by using -Djit=true

jgmdev opened the pull request:
Instead of having a different branch with luajit support this PR adds a new build option `jit` which when set to true will build against luajit. Added official compatibility layer from https://github.com/keplerproject/lua-compat-5.3 with minor path adaptations instead of using the 5.2 adaptation from franko, and all of this is only installed/used if lite-xl was built with `-Djit=true`

![2022-03-13_20:50:09](https://user-images.githubusercontent.com/1702572/158087456-ba557466-4050-495e-b3bb-933f45ffd2e2.png)

I included @franko fixes for stuff like `common.utf8_chars()`, macos files, and not sure if missed something, requesting review!

takase1121 commented:
So now we've moved on from having 1 version of lua (5.2) to 3 (luaJIT/5.1, 5.2, 5.4)! Yay!

Fine, exclude 5.2. We still have 5.1 and 5.4. So what is it? Is it 5.1 or 5.4? 

Can I use my "to be closed" (lol the syntax is bad anyway) variables or not?

Sure, you can write a gazillion polyfills to bridge each other, but what should I (plugin dev) use? 5.1? 5.4? Mixture of both? JavaScript?

jgmdev commented:
> What's the point of having a lingua franca if everyone speaks their own dialect?

so far is working with all plugins I have installed, the performance gains for some operations are nice and it is an optional build time flag which isn't default but nice to have. It seems co-routines are 3x faster than in regular lua and we are abusing them so it feels smoother.

Guldoman commented:
Do we *really* need this?
The amount of code this PR adds is a bit ridiculous. Just looking at the C side, this adds ~4k lines, while our current C code totals at ~3k lines.

I understand that LuaJIT improves performance, but that doesn't justify this added complexity.

There is also the problem that this targets Lua 5.3 while we target 5.4; I don't know how much that could affect things, tho.

jgmdev commented:
> There is also the problem that this targets Lua 5.3 while we target 5.4; I don't know how much that could affect things, tho.

Yes, I searched the keplerproject to see if they had a 5.4 compat lib but that was the most recent one that expands luajit too, well I guess this will be moving to a branch of its own as it has been for now :P

Jan200101 commented on subprojects/luajit.wrap:
```diff
@@ -0,0 +1,5 @@
+[wrap-git]
+directory = luajit
+url = https://github.com/franko/luajit
```
personally, I think it would be better if we moved the wrap to a more indepdendent source, like wrapdb.

Jan200101 replied:
At the moment, wrapdb does not distribute LuaJIT.
If this PR goes forward, feel free to inform me so I can contribute one.

jgmdev replied:
Just rebased this branch to latest master with latest dirmonitor changes, if you could do the luajit wrapdb would be nice, even if it is not desirable to add this change to master maybe we can add a master-jit branch that could be kept in sync with master since most of the changes on this branch aren't too invasive. Who knows if in some more years luajit will catch up or a compatibility lib for 5.4 ...

Jan200101 replied:
good news: the luajit wrap is done

bad news: luajit has a terrible directory structure that assumes in-tree builds and thus prevents it from actually building.

I'll need to talk to some people to figure out how to deal with this

franko replied:
I made a LuaJIT repository with a Meson build that is maintained and used by a (few) other people. It works on MacOS, Windows and Linux without known problems.

I sometimes have the impression that some people just want to remove all my personal contributions from this project.

Jan200101 replied:
> I made a LuaJIT repository with a Meson build that is maintained and used by a (few) other people. It works on MacOS, Windows and Linux without known problems.

Your repo has various modifications from the original LuaJIT source code and the unclean merges have left it in a state where it is hard to reason what was really changed.

Using your meson.build ontop of the official LuaJIT source code is impossible since you changed a few headers (maybe even more) as well as abstracted part of the build process into its own executable, which I doubt would ever be upstreamed.

Its not about not wanting to use your version, it is about using `LuaJIT`, not your flavor of it

> I sometimes have the impression that some people just want to remove all my personal contributions from this project.

If you really want to spin it that way, sure, we've been slowly but surely removing a bunch of code and dependencies added by you as well as moved the build process to depend on a more decentralized method of building.

But that is not what this is about, to me at least.
As previously mentioned, you made changes to LuaJIT which are hard to track and thus it is unknown how far the official version and your version diverge.
If we continue going down that route we will end up with a specialized fork with none of the changes ever upstreamed and upgrading turning into a tedious process of whack-a-diff, moving to using the upstream source code with a patch applied to supply the build system stuff is the better option that will improve the update process as well as allow us to maintain the build system stuff independently of LuaJIT.

If you went missing tomorrow we would need to do this anyway, since you are the only one with control over LuaJIT repo and we really don't need that point of failure.

franko commented:
Too much complexity here, I do not agree with this change even if it may be technically sound.

I would like to switch back to Lua 5.2 or even 5.1 and have the jit option with low engineering complexity.

Luajit is naturally compatible with Lua 5.1 and to some extent with Lua 5.2.

Lua 5.4 brings no added value on the table while luajit does.

We need to maintain engineering complexity low while providing added value and with Lua 5.4 we are in the wrong direction.

Either we support Lua 5.4 only or we support luajit using a compatible Lua version.

Jan200101 commented:
Bringing some Discussion from Discord over here.

LuaJIT has switched to a rolling release distribution model and thus there will be no versioned releases in the (at least recent) future. (see https://github.com/LuaJIT/LuaJIT/issues/665#issuecomment-784452583).

Thus if we decide to go with that we are either stuck having to directly or indirectly (wraps) vendor a LuaJIT version which is likely to quickly go out of date.

or use the last tagged released in 2017 (for reference, 5.2 released on 16 Dec 2011 and 5.1 released on 21 Feb 2006).

Relying on distributions would be unwise because they claim to distribute the latest tagged release but many still maintain out of date patches to the latest commit at the time.
Taking the Fedora RPM as an example, for the longest time they patched LuaJIT with many unofficial patches from forks adding new functionality and now they maintain a, at this point, 6 month out of date patch for the 2.1 branch.



Going one further, we would also need to consider if we are capable of supporting Lua 5.1 or 5.2 on our own.
All the releases except 5.4 are considered EOL

franko commented:
LuaJIT provides stable releases in the 2.0 branch and dev beta releases in the 2.1 branch. Both of them are actively maintained and both of the support the same language, Lua 5.1 with some 5.2 extension. We can let meson find the distribution provided luajit and fallback to our subproject. I don't see the problem here. If at some point Lua 5.1 or 5.2 will no longer be packaged we have our subproject. Lua is thought to be embedded in larger projects so that's the way it is supposed to be used.

As for PUC-Rio Lua, each major version is feature-complete, battle-tested and perfectly usable. None of them is end of line. If you upgrade from one Lua major release, for example from Lua 5.1 to Lua 5.2 is just because you want to use the new language features, not because 5.1 is no longer fine. For these reasons I don't see any reason we cannot stick to 5.1 or 5.2 for the next twenty years, without any problem.

For Lite XL I don't see any tangible gain moving from 5.1 to 5.4. Maybe the C API evolved a little bit and are somewhat more convenient to use but that's not a big deal.

jgmdev commented:
Just for reference this [wiki page](http://wiki.luajit.org/NYI) list some stuff that is available on luajit 2.1 up to lua 5.3, even if minor it is good to know. 

Also, from the packaging side at least on ArchLinuxARM and ArchLinux they are distributing recent commit from the 2.1 branch without any patches and it is working nicely on both AArch64 and x86_64. Sadly, not all linux distributions seem to make the same sane decision that arch packagers did. In any case, compiling against the 2.1 branch seems like the best option since it has many fixes for different architectures like arm64 where older branches used to crash but 2.1 works perfectly so this is a no brainer unless politics are a thing which results in things not getting done and slowing progress down...

From my personal opinion, I opened this PR since I wished to be able to easily compile master against luajit and test the performance (I love performance what can I do?). I saw various gains like the tokenizer been faster, multithreading with #830 been around 2.5x faster, coroutines execution and file operations been faster, etc... The only issue I found was mentioned on #881 which was weak reference coroutines not getting properly released (which should be easily fixable) and native plugins not properly loading which I fixed on a second commit. So from a technical aspect luajit brings some advantages on the performance department as it should. Not using it in my opinion would be a matter of politics or a chase for the biggest version number (in this case 5.4) because bigger seems to be better or shiny? To be fair Lua 5.4 does brings some advantages like binary operators but that is the only thing I can recall right now. How many years so many projects depended solely on C99 and we are already on 2022, some lag will not hurt much xD

Also, I added the official 5.3 compatibility layer to try to satisfy the appetite of those wanting more recent stuff (including mine even If I end not using it :smile:) and because it seemed like a no brainer. It is the easiest way to get luajit up to lua v5.3 compatibility without us having to do our own thing and instead relying on work from other developers who may know better than us, so I don't see how this over complicates things up. Maybe, instead of adding the whole compatibility layer to repo one can link to it but in the end the results would be the same and it would be mostly a matter of politics and I say mostly because maybe linking directly to the 5.3 compat layer could get us latest fixes at build time which would be a gain unless we link to a specific commit.

So the question is, are we contributors discussing over politics or the technical benefits that one solution may bring over another one? Are we choosing to be zealous or pragmatic? Which one is the most beneficial for all of us in the long run? It seems that the decision for only supporting latest lua 5.4 was done without taking a deep look (benchmarking, performance measuring the various lite-xl components, etc...) on the technical and performance benefits that luajit add, most of the game development industry that depends on lua has chosen to stick with luajit and that surely is meaningful. It is not the same to open the editor and say: "it feels fast", than measuring with numbers the difference of one solution and another using different kinds of equipment, and on the things I have measured there are gains on the discarded solution.

Just my 100 cents opinion :smile: we should all focus on the real gains and forget about politics or being zealous, that would end with the opinion conflicts I have been experimenting/seeing with project contributors that are not going to get us anywhere if we want to benefit from the contributions of everyone we also have to care for other people needs as long as they are mutually beneficial.

franko commented:
Quite agree with your well informed comments but in practice what do you propose? You are @jgmdev a major contributor with @adamharrison and your advice matters a lot. Don't be afraid to be bold.
 I want luajit and I want to drop Lua 5.4 to choose the version that matches better with luajit.

To be clear, in case of discussion you @jgmdev  and @adamharrison have the right to decide over these kind of discussions.

Guldoman commented:
> I would like to switch back to Lua 5.2 or even 5.1 and have the jit option with low engineering complexity.
>
> Luajit is naturally compatible with Lua 5.1 and to some extent with Lua 5.2.
>
> Lua 5.4 brings no added value on the table while luajit does.
>

I **completely** disagree.
This was discussed in #556, and all the main arguments still stand:
- Better releases status
- Basically every distro ships 5.4

I'd also add:
- Improvements to utf-8 compatibility (which are fundamental for a text editor).
- `string.pack` and `string.unpack` are very useful.
- Bitwise operators can speed up the tokenizer.
- Garbage collector improvements.

> We need to maintain engineering complexity low while providing added value and with Lua 5.4 we are in the wrong direction.

Adding compatibility layers on top of luajit doesn't seem the way to go either.

Also going back to 5.2 (or 5.1 even) now, while the transition to 5.4 was done 3 months ago (three!) is quite counterproductive.

> We can let meson find the distribution provided luajit and fallback to our subproject. I don't see the problem here. If at some point Lua 5.1 or 5.2 will no longer be packaged we have our subproject.

The obvious problem here is having to maintain that ourselves.
Who's going to do that? Is there anyone that can ensure us that in 3 months they'll still be here to maintain them? What about 6 months? A year?
Lua 5.4 is actively maintained and distros ship it. There's really no reason to switch away from that.

> From my personal opinion, I opened this PR since I wished to be able to easily compile master against luajit and test the performance (I love performance what can I do?).
> I saw various gains like the tokenizer been faster, multithreading with #830 been around 2.5x faster, coroutines execution and file operations been faster, etc...

How much faster was it? How was it tested?

> The only issue I found was mentioned on #881 which was weak reference coroutines not getting properly released (which should be easily fixable)

To be fair the garbage collector not working correctly is a major red flag for me.
And the fix produced was just an ugly workaround.

> Not using it in my opinion would be a matter of politics or a chase for the biggest version number (in this case 5.4) because bigger seems to be better or shiny? To be fair Lua 5.4 does brings some advantages like binary operators but that is the only thing I can recall right now. How many years so many projects depended solely on C99 and we are already on 2022, some lag will not hurt much xD

This is a really bad take.
The main reason we moved to 5.4 was for availability on modern distributions and this hasn't changed.
Also, using a modern version ensures that we're still supported, and allows us to more easily upgrade to future Lua releases which could bring even more improvements.

adamharrison commented:
So, I'd prefer we stick with 5.4 for the following reasons:

We've discussed this many times: on discord, here: https://github.com/lite-xl/lite-xl/discussions/556, https://github.com/lite-xl/lite-xl/pull/781, etc... The conclusion was generally to move to 5.4 (and Franko, I believe you gave this your blessing at the time). We can always discuss going back on that; but I feel like there should be a good, uncontroversial reason to do so. (i.e. it's *a lot* more performant, or it's *a lot* simpler).

To start off, the main lua branch is generally simpler to compile, and more broadly supported. No special tweaking is necessary, no extra wraps required; we simply pull from the official github repository, and are done with it. This is quite attractive. As Jan mentioned previously in this thread, with luajit, you had to maintain a fork of the repository which had its own modified build process that Jan felt difficult to compare with upstream. That isn't ideal.

With regards to lua distributed via the normal operating system package managers, I believe we did determine that this was about even between JIT and 5.4. That said; the state of JIT in each OS doesn't sound good based on Jan's post. That kind of kludging is never good, and I worry about that creating issues depending on exactly how each OS is patching and updating it, since luajit no longer does official releases. If we simply use PUC Lua5.4, we're automatically guaranteed to be pretty standard on every OS package manager, which will likely lead to many less headaches. We're also guaranteed to be closer to any future lua releases that we may want upgrade to in future.

Another benefit is that there's no additional hoops to jump through WRT to distribution. For example; we've had several people try and grab the jit build for mac, only to have it get SIGKILL'd by the OS because it wasn't signed appropriately (because Mac requires us to sign the package if we're going to have JIT-enabled lua). If we don't use JIT, that whole problem, and the complexity associated just disappears; lua is just included in the project and just runs normally without any fancy syscalls or black magic.

In 5.4, we do get access to a number of nice things out of the box (like letting our nested syntax stack go deeper, for example, and as jgmdev mentioned, bitwise operators, and other goodies). If PUC Lua continues to evolve, we'll have the advantage of evolving with it.

The only advantage I see with JIT, is a potential speedup.

However, in practice, while a "drop-in" speedup is nice; much more significant speedups would come from re-architecting existing systems to be more performant; optimising at the macro level, rather than the micro level; or if it's particularly egregious, simply moving it into C (which is generally at least an order of magnitude more performant than even LuaJIT). 

Consider for example, the recent fix involving simply changing the order of rules in a syntax highlighter that resulted in orders of magnitude speed increases. 

Sure it's nice for certain routines to run 2x as fast, or 1.5x as fast; but frequently, in large programs like lite-xl, the overall performance gain in general isn't going to be huge; and, isn't even necessarily the thing I think we should be aiming for, over and above every other concern. 

That said; even if you *do* want additional micro-performance, 5.4 does tend to be more performant than 5.1 (somewhere in the area of 2x in micro-benchmarks). So if we do decide to go with 5.1 over 5.4, anyone who doesn't use JIT, for whatever reason, is going to be stuck with worse performance compared to if we had upgraded. 

Not all performance is created equal; either. For it to really count, it has to be noticeable for the user. In that sense; I think some benchmarks are good, and others not so good. Just having everything humming along a smidge isn't necessarily even going to be something anyone notices. If we're really concerned about performance, I think we'd probably get better results from simply identifying where these pain points are (scrolling, for example, is a huge pain point; by far and away the largest complaint), and re-architecting the systems surrounding them, rather than simply hoping JIT, with its aforementioned drawbacks, fixes the problem enough that we don't need to worry about it.

jgmdev commented:
> To be clear, in case of discussion you @jgmdev and @adamharrison have the right to decide over these kind of discussions.

Besides the inital process api implementation (which was rewritten like 2 times to drop reproc), some docs and lsp plugin I don't consider my self a huge contributor. Also I don't think the subset of contributors to decide is that small, @Guldoman  and others like @Jan200101 have also done nice contributions, Guldo has constantly helped me troubleshoot issues even if we disagree on subjects like this one :wink: that is where maturity shines, when having different views and we still search for mutual benefits we can agree on, understanding that both parts have to collaborate for that to work. Also besides us been contributors we are also part of the community and the community that uses and test the editor should also have a say if they are properly informed.

Anyways, If my opinion has any value: 

* knowing there is a compatibility layer that works for luajit that targets lua v5.3 (which we don't have to maintain) and that works with current code base with minor changes
* there is people who desire luajit like my self basically for the performance gains
* we would also get ffi for free which reduces the amount of code that needs to be compiled to use a native library
* we would loose bitwise operators from 5.4 but they aren't used on all plugins I have tested or the core and the bits library has been providing that functionality just fine so far (we could also benchmark the different but this is rarely used)
* there is no known major plugin that is making use of lua 5.4 specific features
* with luajit we still get large integers support

I would propose to target a max version of lua 5.3 for core and plugin developers sanity sake. This would allow us to keep luajit support while still having a decent enough lua target version that is 2 versions above 5.1 The compatibility layer can be enabled when a user is compiling against a lua version < 5.3 or luajit. 

That is my proposal, details on how to integrate the compat layer are minor in my opinion (I don't consider my self a code purist, as long as it is well architected, documented and performant everything else is just eye candy imho). The most important feature for most users of lite-xl (me included) is how performant it is in contrast to other code editors and making the proposed change would for the most part improve that aspect not the other way around. (there are other editors I could switch to like Sublime which is also good on the performance department but I'm not a python fan)

Another optional proposal would be to write a plugin (if possible) that could serve to benchmark the different components of the editor, which should be tested on different equipments if possible, to see how much they are impacted by the lua core been used. This would provide solid evidence and not only a hunch or placebo effect of the gains or losses.

decalek commented:
Since bit-ops were mentioned several times in the thread as an advantage of PUC 5.4 over LuaJIT above??? - LuaJIT 2 has build-in, [compiled to machine code](http://wiki.luajit.org/NYI#libraries_bit-library) bit ops from the day one. As with 5.3 integers, PUC Lua is the follower here.

About the other mentioned obstacle - the state of the Linux packaging: Actually, the last problematic package is the Debian one, all other major distros are close enough to the upstream. The mentioned Fedora package (which I am using unmodified with Lite XL and Fedora COPR provided `neovim`) is also close enough, and @Jan200101, being in the Fedora packaging group, could easily become co-maintainer and help to improve the package up to the standards himself.

It is time for us (the LuaJIT community) to sort out the Linux packaging issue anyway - the Debian packaging (mostly because it not defines `LUAJIT_ENABLE_LUA52COMPAT`) imposes problems to many other projects e.g. Fennel (ton of workarounds for support exact 5.1 level of the language, e.g. because the missing __pairs/__ipairs).

About the Apple issue - don't worry! This religion has enough fanatic followers - Someone will volunteer for the organizational burden eventually, once tired to run with `jit.off` and be slower than the more mortals on his so shiny device :-)

As @jgmdev mentioned above, with LuaJIT FFI, `ljsyscall`, `luapower`, etc. the plugins authors will have the freedom to do crazy things :-)

`tree-sitter` brought for the editors the comfort of accurate highlighting, because of its full **incremental parsing** support. code-mirror 6 followed the suit with the built-in Lezer. I believe, that we are able to reach the same level - incremental parsing, building on top of on [LPegLJ](https://github.com/sacek/LPegLJ#differences-from-lpeg-v10) too (trough artificial reconstructing of the LPEG VM state before the diff, as function of compact, byte encoded CST, akin to Lezer's). PEG grammars are easier for writing (than LR), LPegLJ supports left recursion and we could even translate ANTLR and tree-sitter ones to PEG semi-automatically (the opposite direction PEG->LR is hard, AFAIK). Furthermore, PEG allows easy grammars nesting which will be **clean advantage** over tree-sitter - e.g. see this [C11 parser](https://github.com/edubart/lpegrex/blob/main/parsers/c11.lua).

Having said all this, I would like to propose to the project owner and the maintainers team the following approach:
 * Stay on PUC 5.4 at least for the 2022 releases. I personally will build and adapt to LuaJIT my own instance, using the branch of the current PR and will try to be of service to @jgmdev, with its support. If everything works - we may start to do binary builds explicitly marked as _experimental_.
 * Bring the matter with 5.4/LuaJIT2.1 compatibility layer to broader LuaJIT SIG, with the proposal about joint-forces automatic (and optimizing) compilation of 5.4 syntax to LuaJIT (source or bytecode) on the fly, with a new package loader,  in the style of MetaLua. @franko is used to this sort of wizardry anyway ;-).
 * When we are ready with fast enough 5.4/5.5? transpiler (testing proves that code transformation doesn't lead to startup slowdowns), switch releases to LuaJIT 2.1, with the option for the plugin authors to mark their sources as LuaJIT, e.g. with comment magic like current mod-version, when they prefer pre-ENV style LuaJIT syntax, in which they are perhaps used to, authoring for neovim or the other popular platforms.

Jan200101 commented:
I do not wish to partake in this conversation much longer but I do lay out some things.

> jgmdev:
>  To be fair Lua 5.4 does brings some advantages like binary operators but that is the only thing I can recall right now.

The added benefit of using Lua 5.4 is upstream support.
Any bugs, vulnerabilities, etc. can all be reported to the (PUC-Rio) Lua developers and fixed in any following releases.
With Lua 5.1 or 5.2 this is not possible and the continued developed would fall on us

At this time this point seems moot since anything Lite-XL will do is sanctioned by the user but this is a matter of future proofing for when the project becomes capable of more either out of the box or using [plugins](https://github.com/lite-xl/lite-xl-plugins/pull/50)

> Franko:
>  I want luajit and I want to drop Lua 5.4 to choose the version that matches better with luajit.

independent of what route we are going it is fact that distributing a lua and a luajit release has caused confusion among users on which version to choose.

Polyfilling the incompatibilities between the normal and luajit versions is possible, but the number of contributors is numbered and maintaining multiple release variants seems fruitless from past attempts.

> decalek:
>  Since bit-ops were mentioned several times in the thread as an advantage of PUC 5.4 over LuaJIT above??? - LuaJIT 2 has build-in, compiled to machine code bit ops from the day one. As with 5.3 integers, PUC Lua is the follower here.

Lua 5.2 implements these too (https://www.lua.org/manual/5.2/manual.html#6.7) and with 5.3 the switch was made from bitwise functions to bitwise operators

Lite-XL introduced a polyfill for it in the Lua 5.4 merge request https://github.com/lite-xl/lite-xl/blob/16fcb2e7512cbae16de133d249fda8d3ce092ef8/data/core/bit.lua though it is limited to 32 bits because of the constrains of the original implementation

> and @Jan200101, being in the Fedora packaging group, could easily become co-maintainer and help to improve the package up to the standards himself.

As much as I would love to, I cannot see how.

Making releases for a project that has no releases and is entirely rolling release is not impossible but keeping up with upstream would quickly become tedious requiring me to use the spare time I have to do anything to dissect LuaJIT, ensure everything continues working and then make frequent builds.

The only sustainable solutions I see would be to either vendor LuaJIT (would still go out of date but the version we have would be confirmed to work against Lite-XL) or patch a meson.build in using wraps (if the build system changes this would break horribly)

panekj commented:
> Making releases for a project that has no releases and is entirely rolling release is not impossible but keeping up with upstream would quickly become tedious requiring me to use the spare time I have to do anything to dissect LuaJIT, ensure everything continues working and then make frequent builds.

There is [OpenResty's LuaJit](https://github.com/openresty/luajit2) which regularly tags repo. It's also a version of LuaJit to which we switched in Alpine Linux.

franko commented:
In response to @adamharrison: yes I gave my blessing to the decision to go with Lua 5.4. I still maintain that we can stay with Lua 5.4 but without LuaJIT.

The compatibility layer I see here is too much code, this is already a technical debt that is not justified.

Otherwise I would have preferred to not move to Lua 5.4 and to keep LuaJIT bu that's okay, I accepted the decision taken by @adamharrison to move to Lua 5.4.

Otherwise, for you to know: I don't buy all the arguments that Lua 5.2 is end of line or that LuaJIT is not viable. Please Jan don't reiterate again your arguments. Lua 5.2 and LuaJIT 2.1 or 2.0 are both perfectly viable for me for this project.

I care more about keeping the engineering complexity and the technical debt of the project as low as possible and maximize the useful features we offer to the user. In this optic going to Lua 5.4 and forgot LuaJIT is perfectly fine so I gave my consent.

I cringed a little bit when you did the native plugin layer @adamharrison because that was unneeded complexity to my eyes but I closed an eye on this.

So I propose to reject this PR despite how much I appreciate the amazing work of @jgmdev and the fact that I would love to have luajit myself. We can continue to keep luajit available on the 2.0 branch. I am still using Lite XL 2.0 as my daily driver and many Lite XL users in the wild are using 2.0.

I have the impression that for the 2.1 you are messing doing more bike-shedding (or yak-shaving, I don't know...) than what is needed but I am waiting patiently to se this new release. I also didn't appreciate removing all the dir monitoring code I finely tuned and worked perfectly fine as far as I can say. We are burning too much energy in bike-shedding, rewriting things that works well instead of adding new useful features for the users. The only useful feature in 2.1 I am aware of is the line wrapping mode but may be I am missing other things. I care only about useful features not about if we use wrapdb or not or if we use Lua 5.2 or 5.4.

PLEASE DO NOT REPLY TO THIS MESSAGE TO FURTHER DISCUSS: everyone got to express his/her point of view.

REPLY ONLY if you have something short and factual to say. I leave the final word for this PR to @adamharrison, I gave my advice and other people here as well. We may also leave this branch open if we decide so, no problem on my side but I think it is better to decide in one sense or the other.

@jgmdev, as this is your PR you are free to make new comments as you like.

adamharrison commented:
> REPLY ONLY if you have something short and factual to say. 

Yeah; just wanted to clarify:

> I cringed a little bit when you did the native plugin layer @adamharrison because that was unneeded complexity to my eyes but I closed an eye on this.

Don't worry about this; this was always a kludge, and I agree it's a bit overwrought. The proposed solution for this is to simply use `-rdynamic` and strip out the compatibility layer. We've discussed that, and I think that's the better solution, if it's acceptable. But we do need some way of interfacing natively with lua, if we are going to do static builds. But we can talk about it in another thread.

> The only useful feature in 2.1 I am aware of is the line wrapping mode

I've actually written the changelog on that as well; please feel free to review it; suffice to say, there is significantly more than just linewrapping.

jgmdev commented:
Ok, i will try to keep this branch in sync with master by rebasing from time to time for those interested until master has enough incompatible code where it is no longer viable. Also will later report to luajit the weak reference issue or tackle it directly on this branch.

adamharrison commented:
> Ok, i will try to keep this branch in sync with master by rebasing from time to time for those interested until master has enough incompatible code where it is no longer viable. Also will later report to luajit the weak reference issue or tackle it directly on this branch.

If it is up to me, then I actually do kinda like the suggestion of @decalek . If the wider LuaJIT community can accommodate newer lua releases, with a transparent transpiler, or what have you, then that should be easy to slot in to the build process, just as it was when we were on 5.2, without any additional engineering complexity. If it is fully compatible, there should be no comparable difference in how the code executes (at least, in the vast majority of cases). I'd still want to have the option to enable the JIT or not, rather than making it the default, but I don't see why that wouldn't work.

takase1121 commented:
That's the problem - we cannot just let something that are basically
incompatible with each other "optional". Either we switch back (ugh) to
luaJIT or use 5.4, there shouldn't be a choice where it shouldn't.

On Mon, 14 Mar 2022, 14:07 Jefferson González, ***@***.***>
wrote:

> What's the point of having a lingua franca if everyone speaks their own
> dialect?
>
> so far is working with all plugins I have installed, the performance gains
> for some operations are nice and it is an optional build time flag which
> isn't default but nice to have. It seems co-routines are 3x faster than in
> regular lua and we are abusing them so it feels smoother.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/lite-xl/lite-xl/pull/880#issuecomment-1066414019>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE6UHTHYZJLXMRGUX3D5IJLU73JSDANCNFSM5QUDEJKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

mnemnion commented:
Here to offer some outside perspective.

The Lua community is small, and the split between LuaJIT and PUC Lua is complete.  

A project which wants to grow its contributors would be wise to make the LuaJIT 'dialect' standard for core contributions, and encouraged for plugins.  With leafo's `setfenv` and `getfenv` implementations, this has plenty of expressivity. 

LuaJIT and PUC Lua isn't Python 2.7 and Python 3.n.  There are new users of LuaJIT every month, projects which use it have no intention of changing that, ever. 

It's just a suggestion. I have a large and growing collection of code which works with LuaJIT, so lite-XL is only interesting to me if I can build it with that engine, but that alone shouldn't affect your deliberations.

I will observe that a community fork which deliberately removes LuaJIT compatibility has perhaps made a hasty decision in that regard, if the goal is to steward the existing code while adding features.  Building with LuaJIT is in fact a feature, for me, the most important one.

adamharrison commented:
Tentatively closed for now; this functionality is available in https://github.com/pragtical/pragtical.
